File: PYTHON/quick_sort.py


def partition(top, end, array):

	pivot_index = top
	pivot = array[pivot_index]
	

	while top < end:
		
	
		while top < len(array) and array[top] <= pivot:
			top += 1
			
		
		while array[end] > pivot:
			end -= 1
		
		
		if(top < end):
			array[top], array[end] = array[end], array[top]
	

	array[end], array[pivot_index] = array[pivot_index], array[end]
	

	return end
	

def quick_sort(top, end, array):
	
	if (top < end):
		
	
		p = partition(top, end, array)
		
	
		quick_sort(top, p - 1, array)
		quick_sort(p + 1, end, array)
		

array = [ 10, 7, 8, 9, 1, 5 ]
quick_sort(0, len(array) - 1, array)

print(f'Sorted array: {array}')
	

File: JAVA/Sorting/quickSort.java
import java.util.Scanner;

public class quickSort {
	
public static void main(String[] args) {
	Scanner scanner = new Scanner(System.in);
    int n;
    System.out.print("Enter the number of elements in the array: ");
    n = scanner.nextInt();
    int[] arr = new int[n + 1];
    System.out.print("Enter the array elements: ");
    for (int i = 0; i < n; i++) {
		arr[i] = scanner.nextInt();
    }
    arr[n] = Integer.MAX_VALUE;
    quickSort(arr, 0, arr.length - 1);
	System.out.println("Sorted Array: ");
    for (int i = 0; i < arr.length - 1; i++)
    System.out.print(arr[i] + " ");
}

public static void quickSort(int[] arr, int l, int r) {
	if (l < r)
    {
        int j = partition(arr, l, r);
        quickSort(arr, l, j);
        quickSort(arr, j + 1, r);
    }
}

public static int partition(int[] arr, int l, int r) {
    int pivot = arr[l];
    int i = l, j = r;
    while (i < j) {
        do {
            i++;
        } while (arr[i] <= pivot);
        do {
            j--;
        } while (arr[j] >  pivot);
        if (i < j) {
            swap(i, j, arr);
        }
    }
    swap(l, j, arr);
    return j;
}

public static void swap(int i, int j, int[] arr) {
    int temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;
}
}